Remove all uses of `not_expecting` in the parser.

fpottier opened the pull request:
The grammar is tweaked in three places so as to remove the need to use `not_expecting`.
Indeed, it is desirable for the grammar to reflect *exactly* the intended language,
as opposed to a larger language than intended.
This should make it easier in the future to give good syntax error messages.

I note that we might consider deprecating `Syntaxerr.Not_expecting`.
Unfortunately, it is part of `compiler-libs`, so possibly used outside the compiler.

fpottier committed:
Make the grammar more precise by explicitly disallowing `NONREC`
inside `str_type_extension` and `sig_type_extension`. This removes
the need for post-hoc checks, and removes two calls to `not_expecting`.

fpottier committed:
Remove the auxiliary function `not_expecting`, which is no longer used.

fpottier committed:
Parser: A simpler definition of `ext_attributes`.
(Not sure why a complicated definition was used.)

fpottier committed:
Inline `ext_attributes` into `let_binding`.

fpottier committed:
Parser: parameterize `let_binding` and `let_bindings` with `EXT`,
which can be instantiated with either `ext` of `no_ext`.
This allows getting rid of one use of `not_expecting`
in the auxiliary function `class_of_let_bindings`.

fpottier committed:
Parser: remove the production "expr: UNDERSCORE",
whose semantic action calls `not_expecting`.
Such a production is bad style and tricks Menhir
into believing that an underscore is a valid expression.

Drup commented:
> I note that we might consider deprecating `Syntaxerr.Not_expecting`.
> Unfortunately, it is part of `compiler-libs`, so possibly used outside the compiler.

Just to react to that bit: It is official acceptable to break compat in compiler-libs; especially in such a place which is highly unlikely to be used independently.

fpottier commented:
@Drup: true, but I suppose it is still preferable not to break anyone's code.
A Google search suggests that Merlin (for instance) uses `Not_expecting`.

fpottier committed:
Parser: remove several uses of `syntax_error()` by explicitly using
`no_override_flag` instead of `override_flag` when this flag is disallowed.
A conflict is avoided by decorating `override_flag` with `%inline`.

fpottier commented:
Pushed one more commit, which removes a few uses of `syntax_error()`.
This aspect of the parser is now clean, I think.
There are no more places where syntax errors are detected in the semantic actions.

fpottier commented:
Travis seems to complain that this PR does not come with tests.
I don't think any new tests are required.
All existing tests pass, and the ASTs are unchanged.
The only change in behavior is that the messages "wildcard not expected" or "nonrec not expected" which could arise in some situations are transformed into plain "syntax error" messages.

Octachron commented:
The "new tests" check is an allowed failure; however Travis is also complaining about the long lines in `parser.mly` ( line 1404, 1418, and 2007) and the lack of change entry. Concerning the last point, you can probably add this PR to the list of satellite PRs [here](https://github.com/ocaml/ocaml/blob/trunk/Changes#L341) .

fpottier committed:
Eliminate long lines.

fpottier committed:
Mention this PR in the change log.

fpottier committed:
Parser: introduce `virtual_with_mutable_flag` and `virtual_with_private_flag`
in order to eliminate a redundancy between productions in two places.
This removes two TODO markers.

fpottier committed:
Run `make promote-menhir`.

fpottier commented:
OK, I have removed the lines over 80 characters long, mentioned this PR in the file `Changes`, and taken this opportunity to add another tweak which improves sharing between productions and removes a couple TODO markers.

gasche commented:
(I haven't had a chance to look at this yet, sorry.)

trefis commented:
> A Google search suggests that Merlin (for instance) uses `Not_expecting`.

I don't think merlin is relevant in this case:
- we don't use compiler-libs but a modified version of each version of the compiler frontend
- I don't think we really use the exception extensively, I think the google hits are just a result of our duplication of the code of the compiler
- even if we actually used it, we have a to deal with a ton of tedious changes with each compiler release, I don't think that one makes much difference (or would be particularly tedious to handle)

That said: I only read the discussion, not the implementation, and I didn't check merlin codebase, it's just intuition.

gasche commented on parsing/parser.mly:
```diff
@@ -2011,9 +2011,8 @@ let_bindings:
-      { let (ext, attr) = $2 in
-        mklbs ~loc:$sloc ext $3 (mklb ~loc:$sloc true $4 (attr@$5)) }
+    LET ext = ext attr = attributes rec_flag let_binding_body post_item_attributes
+      { mklbs ~loc:$sloc ext $4 (mklb ~loc:$sloc true $5 (attr@$6)) }
```
Note: instead of shifting numbers, it would be nice to name the symbols (`rec_flag`, `lb_body`, `post_attrs`).

We want to gradually move to readable semantic actions.

(No need to update the PR, this is for the next time.)

gasche commented on parsing/parser.mly:
```diff
@@ -1684,8 +1684,6 @@ expr:
   | expr attribute
       { Exp.attr $1 $2 }
-  | UNDERSCORE
-     { not_expecting $loc($1) "wildcard \"_\"" }
```
Note: this was added by @alainfrisch in 110835872cd75a495c1a58937de5499815842397 to address [MPR#6400](https://caml.inria.fr/mantis/view.php?id=6400), namely the following usability issue with the previous grammar:

```
# [{pstr_desc = Pstr_extension (extension_of_error error, _)}];;
Error: Syntax error: operator expected.
```

gasche replied:
With your current patch applied, the new behavior is as follows:

```
# [{pstr_desc = Pstr_extension (extension_of_error error, _)}];;
                                                          ^
Error: Syntax error: ')' expected
Line 1, characters 29-30: This '(' might be unmatched
```

gasche commented on parsing/parser.mly:
```diff
@@ -2483,22 +2483,22 @@ label_declaration_semi:
+  optional_type_parameters mkrhs(type_longident)
   PLUSEQ private_flag str_extension_constructors post_item_attributes
       { let (ext, attrs) = $2 in
-        if $3 <> Recursive then not_expecting $loc($3) "nonrec flag";
```
Before this change, the error message was as follows:

```
# module type M = sig type nonrec t += A end;;
                           ^^^^^^
Error: Syntax error: nonrec flag not expected.
```


After this change, the error message is as follows:

```
# module type M = sig type nonrec t += A end;;
                                    ^^
Error: Syntax error: 'end' expected
Line 1, characters 16-19: This 'sig' might be unmatched
```

gasche reviewed (commented):
@let-def and myself just reviewed this PR.

Our conclusion is that it does go in the right direction for our work (with @fpottier) on error messages, but it leads to noticeable short-term regressions in the quality of error messages (while testing we also found and reported [MPR#7858](https://caml.inria.fr/mantis/view.php?id=7858)). It is too risky to merge into `trunk` right now, because we have no idea before said error-messages work is going to land in trunk before 4.08.0. So we propose to create a separate branch, `better-error-messages`, to keep track of this work (and other similar improvements that are not trunk-safe).

More details on the big picture:

- We agree that `not_expecting` should go away.

- There are three kind of parse errors:
   - Parses that are representable in the AST but violate the invariants that we have informally specified in the language reference. These should go in our Ast_invariants checker to also apply to ppx-produced ASTs.
   - Things that are not representable in our AST. They have to be refused at parse time, but we agree with @fpottier that Menhir-internal error-reporting mechanisms are much preferable to `not_expecting` in the long run. The grammar (without running the semantic actions) should *never* accept non-AST-representable parses.
   - Some things are parse errors today but in fact it would be conceivable to accept them (for example, extensions in let-bindings in class declarations). In these cases, we should seriously think about extending the reference grammar for coherence and regularity.

fpottier commented:
Thanks for your review! Creating a separate branch sounds like a good idea, as it removes some of the need for premature reviewing. Perhaps it should be called `parser-changes`, as it is not just about getting better error messages but also (first) about cleaning up the parser.

I am aware of some of the regressions (in the syntax error messages) that you point out. In my mind, most of the current syntax error messages are very bad anyway, and it does not matter if the ratio of very bad messages goes up from 99.98% to 99.99%. For instance, look at this:
```
# (3+);;
Error: Syntax error: operator expected.
```

gasche commented:
If we make this a branch in the `ocaml/ocaml` repository, you can still send PRs against it and have them reviewed -- the github interface lets you select which branch a PR is against, with `trunk` the default. (Or you can send PRs against a branch in your own repository, and ask us to review them.) I was not suggesting that you keep a monster branch on your end, to review all those changes at the very end -- I think that would make things harder.
